[WIP] LangGraph-based functional Kaggle solver

YuanmingLeee opened the pull request:
Implement a refactored Kaggle workflow using LangGraph's functional API, providing a modular design for easier readability and maintenance. Each task is encapsulated as a callable node, streamlining collaboration and functionality swapping.

This change improves flexibility and scalability for managing Kaggle competitions while maintaining seamless orchestration.

Closes #314 

#### Demo

1. Download dataset
```
mle bench prepare -c aerial-cactus-identification
```

2. Start vLLM Server
```
vllm serve Qwen/Qwen3-4B-AWQ --enable-auto-tool-choice --tool-call-parser hermes --reasoning-parser deepseek_r1
```

3. Run kaggle competition workflow
```
python ./kaggle_solver.py -c aerial-cactus-identification --model Qwen/Qwen3-4B-AWQ test
```
Demo logs with vllm+qwen3-4B-AWQ model backend:
##### Step1 Setup env, collect basic metadata:
<img width="1694" height="191" alt="image" src="https://github.com/user-attachments/assets/ab1d66c4-4919-45e4-be13-ab97b94f24e5" />

#####  Step2 Advisory Summary:
<img width="1841" height="754" alt="image" src="https://github.com/user-attachments/assets/9afc63a7-c566-463c-857d-a673d08df23e" />

#####  Step3 Planer:
<img width="1848" height="1293" alt="image" src="https://github.com/user-attachments/assets/4e33104b-fadb-4da8-8840-16d659e2b678" />

#####  Step4 Code for each planned tasks sequentially
<img width="1851" height="1204" alt="image" src="https://github.com/user-attachments/assets/b004eb0b-70d8-4250-8f07-e7f63218b44f" />
<img width="1836" height="1719" alt="image" src="https://github.com/user-attachments/assets/f84b8e78-ad95-485e-9670-13ebccd04dbc" />
<img width="1836" height="1866" alt="image" src="https://github.com/user-attachments/assets/f59d084b-b085-4a4f-a9cb-aeb9367343ce" />
<img width="1827" height="1876" alt="image" src="https://github.com/user-attachments/assets/b093dc35-6658-48d0-8660-cd128869beb0" />
<img width="1828" height="1867" alt="image" src="https://github.com/user-attachments/assets/2e878e1d-b979-4478-915f-e0d7d9113192" />
<img width="1822" height="1872" alt="image" src="https://github.com/user-attachments/assets/295f289d-6df9-40ae-baef-b209b98df910" />
<img width="1827" height="1873" alt="image" src="https://github.com/user-attachments/assets/72a1bdc5-c6df-4db7-9a03-27c0f6352d66" />
<img width="1831" height="1870" alt="image" src="https://github.com/user-attachments/assets/d66cbdfa-932a-4fd1-9eed-6b7ed9bdab6e" />
<img width="1828" height="1870" alt="image" src="https://github.com/user-attachments/assets/689c496a-d089-405e-8eef-56840046ee76" />
<img width="1830" height="1873" alt="image" src="https://github.com/user-attachments/assets/d66a4a9e-bb51-486e-a2bd-d246508bcc78" />
<img width="1827" height="1864" alt="image" src="https://github.com/user-attachments/assets/2dfd27b6-8436-4091-9f45-42e510a0014d" />
... (nothing to change for the last step)
Kaggle workflow completed!

#### TODO

- [x] Project setup with a venv (simple sandbox)
- [x] Coder output to a file
- [ ] Debuger actually run the code, rather than just view it, and collect the debugging logs
- [ ] Add plan: output a submission csv

#### Before submitting this PR, please make sure you have:

- [ ] confirmed all checks still pass OR confirm CI build passes.
- [ ] verified that any code or assets from external sources are properly credited in comments and/or in
  the [credit file](https://github.com/MLSysOps/MLE-agent/blob/main/.github/OPEN_SOURCE_LICENSES.md).

YuanmingLeee committed:
:sparkles: [workflow] Add LangGraph-based functional Kaggle solver

Implement a refactored Kaggle workflow using LangGraph's functional API,
providing a modular design for easier readability and maintenance.
Each task is encapsulated as a callable node, streamlining collaboration
and functionality swapping.

This change improves flexibility and scalability for managing Kaggle
competitions while maintaining seamless orchestration.

YuanmingLeee committed:
:hammer: [refactor] Enhance Kaggle workflow with improved state management and error handling

- Fix bug: working dir not found for plan and advisor agents

YuanmingLeee committed:
:hammer: [refactor] Integrate agent classes into Kaggle solver for enhanced task management and debugging

YuanmingLeee committed:
:recycle: [kaggle_solver] Refactor agents and streamline pipeline

Remove redundant agent initialization and workflow cache logic. Simplify the Kaggle solver pipeline by consolidating agent calls and reducing state dependencies.

HuaizhengZhang commented:
Looks good

HuaizhengZhang commented:
`working dir` is all you need 😏

YuanmingLeee committed:
:sparkles: [exp] Add core agents and improve Kaggle competition automation workflow

Introduce `AdviseAgent`, `PlanAgent`, and `CodeAgent` for simplify tasks prompts and LLM calls in the Kaggle competition pipeline.
Enhance utility functions and setup procedures, such as virtual
environment creation, dataset structure visualization, and compatibility checks.

YuanmingLeee commented:
Previous prompt flows were too complex for integration with vLLM, particularly when involving tool calling. 

So I copy and simplified previous agent workflows to exp folder; add some env collection steps for better enforcement with 8B Qwen model (seems like this is the most suitable model for balancing perf and speed with one 4090)

More simplification can be achieved with language graph message passing. 2-3 more days needed for refinement

HuaizhengZhang commented:
> integration with vLLM, particularly when involving tool calling.

I feel chances.

YuanmingLeee committed:
:electric_plug: [agents] Refactor agents to support reusable static methods and state

Refactored `AdviseAgent` and `CodeAgent` to implement agent subgraph calls.

This change improves modularity, simplifies object instantiation, and
enhances maintainability by consolidating logic into reusable components.

YuanmingLeee committed:
:zap: [agents] Refactor tool handling and add dependency management support

- Add utilize `safe_fileio` for secure file I/O within tool handling, and introduced a dependency retrieval method (`deps`) to manage runtime requirements.

Improves code modularity for better aligns with agent processing flows for code generation and execution.

YuanmingLeee committed:
:recycle: [agents] Refactor planning and improve environment handling

- Refactor `PlanAgent` to decouple environment, simplify initialization, and streamline method logic.
- Add `graph` entrypoint to standardize state-based task processing.
- Improve prompts and golden rules compliance in task execution flows.

Enhances maintainability, enforces single-task focus, and aligns planning logic to agent design. Removes redundant platform details for cleaner environment state.

YuanmingLeee committed:
:test_tube: [decorators] Add method_task and method_entrypoint decorators

Introduce `method_task` and `method_entrypoint` decorators to simplify
converting instance methods into reusable langgraph tasks and entrypoints.

This modularizes instance-based task and entrypoint transformation, providing
better code reuse and cleaner abstractions for agents

YuanmingLeee commented:
@HuaizhengZhang A workable example using Qwen3-4B, the coding procedure seems neat and minimal. The debugger agent needs to be added to ensure the code is executable without error and produces submission.csv.

Do not review `exp/decorators.py`; it's for experimental purposes and not yet in effect to simplify the agent code.

YuanmingLeee committed:
:zap: [agent] Add dependency verification and execution logic

- Introduce `verify_code` method to dynamically check and install missing Python dependencies in a virtual environment and execute specified
commands.
- Update `execute_command` function to support additional options for working directory and raw output formatting.

YuanmingLeee committed:
:recycle: [agent] Refactor to integrate `method_task` and `method_entrypoint`

Refactored instance methods to utilize `method_task` and `method_entrypoint` decorators for enhanced readability and encapsulation.

YuanmingLeee committed:
:hammer: [agent] Refactor and enhance method/entrypoint decorators

Refactor `method_task` and `method_entrypoint` to streamline their behavior with instance-specific caching and proper binding. Added overloads for cleaner typing and updated implementation for runtime descriptor management.
